kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,700 @@
+
+  return setmetatable({
+    cache         = cache,
+    stats         = {},
+    hosts         = hosts,
+    r_opts        = r_opts,
+    resolv        = opts._resolv or resolv,
+    valid_ttl     = opts.valid_ttl,
+    error_ttl     = opts.error_ttl or DEFAULT_ERROR_TTL,
+    stale_ttl     = opts.stale_ttl or DEFAULT_STALE_TTL,
+    empty_ttl     = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    search_types  = search_types,
+  }, mt)
+end
+
+
+local function process_answers(self, qname, qtype, answers)
+  local errcode = answers.errcode
+  if errcode then
+    answers.ttl = errcode == NAME_ERROR_CODE and self.empty_ttl or self.error_ttl
+    -- compatible with balancer, which needs this field
+    answers.expire = now() + answers.ttl
+    return answers
+  end
+
+  local processed_answers = {}
+  local cname_answer
+
+  local ttl = self.valid_ttl or 0xffffffff
+
+  for _, answer in ipairs(answers) do
+    answer.name = answer.name:lower()
+
+    if answer.type == TYPE_CNAME then
+      cname_answer = answer   -- use the last one as the real cname
+
+    elseif answer.type == qtype then
+      -- compatible with balancer, see https://github.com/Kong/kong/pull/3088
+      if answer.type == TYPE_AAAA then
+        answer.address = ipv6_bracket(answer.address)
+      elseif answer.type == TYPE_SRV then
+        answer.target = ipv6_bracket(answer.target)
+      end
+
+      table_insert(processed_answers, answer)
+    end
+
+    if self.valid_ttl then
+      answer.ttl = self.valid_ttl
+    else
+      ttl = math_min(ttl, answer.ttl)
+    end
+  end
+
+  if #processed_answers == 0 then
+    if not cname_answer then
+      return {
+        errcode = EMPTY_RECORD_EC,
+        errstr  = EMPTY_RECORD_ESTR,
+        ttl     = self.empty_ttl,
+        -- expire = now() + self.empty_ttl,
+      }
+    end
+
+    table_insert(processed_answers, cname_answer)
+  end
+
+  processed_answers.expire = now() + ttl
+  processed_answers.ttl = ttl
+
+  return processed_answers
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+  local key = name .. ":" .. qtype
+  stats_count(self.stats, key, "query")
+
+  local r, err = resolver:new(self.r_opts)
+  if not r then
+    return nil, "failed to instantiate the resolver: " .. err
+  end
+
+  local options = { additional_section = true, qtype = qtype }
+  local answers, err = r:query(name, options)
+  if r.destroy then
+    r:destroy()
+  end
+
+  if not answers then
+    stats_count(self.stats, key, "query_fail")
+    return nil, "DNS server error: " .. (err or "unknown")
+  end
+
+  answers = process_answers(self, name, qtype, answers)
+
+  stats_count(self.stats, key, answers.errstr and "query_err:" .. answers.errstr
+                                               or "query_succ")
+
+  return answers, nil, answers.ttl
+end
+
+
+local function start_stale_update_task(self, key, name, qtype)
+  stats_count(self.stats, key, "stale")
+
+  timer_at(0, function (premature)
```

Comment:
Should we check if the timer is created successfully?

Reply:
yes, I think we need to check it , if failed, it should report the error

Reply:
resolved in https://github.com/Kong/kong/pull/12305/commits/0ddd237cdc7a0e2e029e11c2f0f5fbf11a4171ba

and I use static function instead of recreating a closure for creating new timers.